include/revng/Pipeline/InspectableLLVMContainer.h

```diff
@@ -0,0 +1,171 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <iterator>
+
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Analysis/ScalarEvolutionExpressions.h"
+#include "llvm/IR/Function.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+#include "revng/Pipeline/ContainerFactory.h"
+#include "revng/Pipeline/ContainerInspector.h"
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Kind.h"
+#include "revng/Pipeline/Step.h"
+#include "revng/Pipeline/Target.h"
+
+namespace Pipeline {
+
+/// Inspector that must be extended to be able to specify how a to deduce
+/// the target associated to a particular llvm global function.
+///
+/// When extented two function must be provided, symbolToTarget must return
+/// the target associated to a function if there is any
+///
+/// compactTargets must collapse the targets into the * target if they are all
+/// presents, do no thing otherwise.
+template<typename LLVMContainer>
+class LLVMContainerInspector : public ContainerInspector<LLVMContainer> {
+public:
+  using StaticContainer = llvm::SmallVector<LLVMContainerInspector *, 4>;
+  LLVMContainerInspector(Kind &K) : ContainerInspector<LLVMContainer>(K) {
+    getRegisteredInspectors().push_back(this);
+  }
+  virtual ~LLVMContainerInspector() = default;
+
+  virtual std::optional<Target>
+  symbolToTarget(const llvm::Function &Symbol) const = 0;
+
+  static bool hasOwner(const llvm::Function &Symbol) {
+    for (const auto &I : getRegisteredInspectors())
+      if (I->symbolToTarget(Symbol))
+        return true;
+    return false;
+  }
+
+  virtual TargetsList
+  compactTargets(const Context &Ctx, TargetsList::List &Targets) const = 0;
```

Comment:
It's kinda weird that `compactTargets` is part of `LLVMContainerInspector` while `expandTargets` is part of `Kind`.
What's the ratio?

Reply:
eh, expand targets is needed in all kinds because it's used always.

compact targets is used just by inspector to collapse the content of a container observed one element at the time into the "*"

i can move compact targets into kinds but it would be often redundant.

Reply:
in the pratical purposes toh they will be provided by the same classes, when one declares a new KindInspector that must provide both